File: cpp/c-api/src/core/gmod_node_metadata.cpp
#include "dnv/vista/sdk/c/core/gmod_node_metadata.h"

#include "../cast_internal.h"
#include "../error_internal.h"

#include <iterator>

using namespace dnv::vista::sdk;
using dnv::vista::sdk::c::toMetadata;

const char* dnv_vista_sdk_gmod_node_metadata_category(const dnv_vista_sdk_gmod_node_metadata_t* metadata)
{
    if (metadata == nullptr)
    {
        c::setLastError("metadata must not be null", DNV_VISTA_SDK_ERROR_INVALID_ARGUMENT);
        return nullptr;
    }

    return toMetadata(metadata)->category().data();
}

const char* dnv_vista_sdk_gmod_node_metadata_type(const dnv_vista_sdk_gmod_node_metadata_t* metadata)
{
    if (metadata == nullptr)
    {
        c::setLastError("metadata must not be null", DNV_VISTA_SDK_ERROR_INVALID_ARGUMENT);
        return nullptr;
    }

    return toMetadata(metadata)->type().data();
}

const char* dnv_vista_sdk_gmod_node_metadata_full_type(const dnv_vista_sdk_gmod_node_metadata_t* metadata)
{
    if (metadata == nullptr)
    {
        c::setLastError("metadata must not be null", DNV_VISTA_SDK_ERROR_INVALID_ARGUMENT);
        return nullptr;
    }

    return toMetadata(metadata)->fullType().data();
}

const char* dnv_vista_sdk_gmod_node_metadata_name(const dnv_vista_sdk_gmod_node_metadata_t* metadata)
{
    if (metadata == nullptr)
    {
        c::setLastError("metadata must not be null", DNV_VISTA_SDK_ERROR_INVALID_ARGUMENT);
        return nullptr;
    }

    return toMetadata(metadata)->name().data();
}

const char* dnv_vista_sdk_gmod_node_metadata_common_name(const dnv_vista_sdk_gmod_node_metadata_t* metadata)
{
    if (metadata == nullptr)
    {
        c::setLastError("metadata must not be null", DNV_VISTA_SDK_ERROR_INVALID_ARGUMENT);
        return nullptr;
    }

    const auto& commonName = toMetadata(metadata)->commonName();
    return commonName.has_value() ? commonName->c_str() : nullptr;
}

const char* dnv_vista_sdk_gmod_node_metadata_definition(const dnv_vista_sdk_gmod_node_metadata_t* metadata)
{
    if (metadata == nullptr)
    {
        c::setLastError("metadata must not be null", DNV_VISTA_SDK_ERROR_INVALID_ARGUMENT);
        return nullptr;
    }

    const auto& definition = toMetadata(metadata)->definition();
    return definition.has_value() ? definition->c_str() : nullptr;
}

const char* dnv_vista_sdk_gmod_node_metadata_common_definition(const dnv_vista_sdk_gmod_node_metadata_t* metadata)
{
    if (metadata == nullptr)
    {
        c::setLastError("metadata must not be null", DNV_VISTA_SDK_ERROR_INVALID_ARGUMENT);
        return nullptr;
    }

    const auto& commonDefinition = toMetadata(metadata)->commonDefinition();
    return commonDefinition.has_value() ? commonDefinition->c_str() : nullptr;
}

int dnv_vista_sdk_gmod_node_metadata_install_substructure(
    const dnv_vista_sdk_gmod_node_metadata_t* metadata, int* outValue)
{
    if (metadata == nullptr || outValue == nullptr)
    {
        c::setLastError("metadata and outValue must not be null", DNV_VISTA_SDK_ERROR_INVALID_ARGUMENT);
        return 0;
    }

    const auto installSubstructure = toMetadata(metadata)->installSubstructure();
    if (!installSubstructure.has_value())
    {
        return 0;
    }

    *outValue = *installSubstructure ? 1 : 0;
    return 1;
}

size_t dnv_vista_sdk_gmod_node_metadata_normal_assignment_name_count(const dnv_vista_sdk_gmod_node_metadata_t* metadata)
{
    if (metadata == nullptr)
    {
        c::setLastError("metadata must not be null", DNV_VISTA_SDK_ERROR_INVALID_ARGUMENT);
        return 0;
    }

    return toMetadata(metadata)->normalAssignmentNames().size();
}

const char* dnv_vista_sdk_gmod_node_metadata_normal_assignment_name_key_at(
    const dnv_vista_sdk_gmod_node_metadata_t* metadata, size_t index)
{
    if (metadata == nullptr)
    {
        c::setLastError("metadata must not be null", DNV_VISTA_SDK_ERROR_INVALID_ARGUMENT);
        return nullptr;
    }

    const auto& names = toMetadata(metadata)->normalAssignmentNames();
    if (index >= names.size())
    {
        c::setLastError("index out of range", DNV_VISTA_SDK_ERROR_OUT_OF_RANGE);
        return nullptr;
    }

    auto it = names.begin();
    std::advance(it, static_cast<std::ptrdiff_t>(index));
    return it->first.c_str();
}

const char* dnv_vista_sdk_gmod_node_metadata_normal_assignment_name_value_at(
    const dnv_vista_sdk_gmod_node_metadata_t* metadata, size_t index)
{
    if (metadata == nullptr)
    {
        c::setLastError("metadata must not be null", DNV_VISTA_SDK_ERROR_INVALID_ARGUMENT);
        return nullptr;
    }

    const auto& names = toMetadata(metadata)->normalAssignmentNames();
    if (index >= names.size())
    {
        c::setLastError("index out of range", DNV_VISTA_SDK_ERROR_OUT_OF_RANGE);
        return nullptr;
    }

    auto it = names.begin();
    std::advance(it, static_cast<std::ptrdiff_t>(index));
    return it->second.c_str();
}

File: bindings/rust/vista-sdk/src/core/gmod_node_metadata.rs
use std::ffi::CStr;

use crate::ffi::core::gmod_node_metadata as ffi;

/// Metadata for a Gmod (Generic Product Model) node.
///
/// Borrowed from the parent [`crate::core::gmod_node::GmodNode`], valid as long as it is.
#[repr(transparent)]
pub struct GmodNodeMetadata(ffi::dnv_vista_sdk_gmod_node_metadata_t);

impl GmodNodeMetadata {
    pub(crate) fn from_ptr<'a>(ptr: *const ffi::dnv_vista_sdk_gmod_node_metadata_t) -> &'a Self {
        assert!(
            !ptr.is_null(),
            "gmod node metadata pointer must not be NULL"
        );
        unsafe { &*(ptr as *const GmodNodeMetadata) }
    }

    fn as_ffi_ptr(&self) -> *const ffi::dnv_vista_sdk_gmod_node_metadata_t {
        self as *const Self as *const ffi::dnv_vista_sdk_gmod_node_metadata_t
    }

    /// Broad classification of this node (e.g. `"ASSET FUNCTION"`).
    pub fn category(&self) -> &str {
        let ptr = unsafe { ffi::dnv_vista_sdk_gmod_node_metadata_category(self.as_ffi_ptr()) };
        assert!(
            !ptr.is_null(),
            "dnv_vista_sdk_gmod_node_metadata_category returned NULL"
        );
        unsafe { CStr::from_ptr(ptr) }
            .to_str()
            .expect("invalid UTF-8 in gmod node metadata category")
    }

    /// Specific type within the category (e.g. `"PROPULSION"`).
    pub fn r#type(&self) -> &str {
        let ptr = unsafe { ffi::dnv_vista_sdk_gmod_node_metadata_type(self.as_ffi_ptr()) };
        assert!(
            !ptr.is_null(),
            "dnv_vista_sdk_gmod_node_metadata_type returned NULL"
        );
        unsafe { CStr::from_ptr(ptr) }
            .to_str()
            .expect("invalid UTF-8 in gmod node metadata type")
    }

    /// Category and type concatenated (e.g. `"ASSET FUNCTION.PROPULSION"`).
    pub fn full_type(&self) -> &str {
        let ptr = unsafe { ffi::dnv_vista_sdk_gmod_node_metadata_full_type(self.as_ffi_ptr()) };
        assert!(
            !ptr.is_null(),
            "dnv_vista_sdk_gmod_node_metadata_full_type returned NULL"
        );
        unsafe { CStr::from_ptr(ptr) }
            .to_str()
            .expect("invalid UTF-8 in gmod node metadata full type")
    }

    /// Human-readable name for this node.
    pub fn name(&self) -> &str {
        let ptr = unsafe { ffi::dnv_vista_sdk_gmod_node_metadata_name(self.as_ffi_ptr()) };
        assert!(
            !ptr.is_null(),
            "dnv_vista_sdk_gmod_node_metadata_name returned NULL"
        );
        unsafe { CStr::from_ptr(ptr) }
            .to_str()
            .expect("invalid UTF-8 in gmod node metadata name")
    }

    /// Common-language name, if defined.
    pub fn common_name(&self) -> Option<&str> {
        let ptr = unsafe { ffi::dnv_vista_sdk_gmod_node_metadata_common_name(self.as_ffi_ptr()) };
        if ptr.is_null() {
            return None;
        }
        Some(
            unsafe { CStr::from_ptr(ptr) }
                .to_str()
                .expect("invalid UTF-8 in gmod node metadata common name"),
        )
    }

    /// Technical definition, if defined.
    pub fn definition(&self) -> Option<&str> {
        let ptr = unsafe { ffi::dnv_vista_sdk_gmod_node_metadata_definition(self.as_ffi_ptr()) };
        if ptr.is_null() {
            return None;
        }
        Some(
            unsafe { CStr::from_ptr(ptr) }
                .to_str()
                .expect("invalid UTF-8 in gmod node metadata definition"),
        )
    }

    /// Common-language definition, if defined.
    pub fn common_definition(&self) -> Option<&str> {
        let ptr =
            unsafe { ffi::dnv_vista_sdk_gmod_node_metadata_common_definition(self.as_ffi_ptr()) };
        if ptr.is_null() {
            return None;
        }
        Some(
            unsafe { CStr::from_ptr(ptr) }
                .to_str()
                .expect("invalid UTF-8 in gmod node metadata common definition"),
        )
    }

    /// Whether this node represents an installation substructure, if the field is present.
    pub fn install_substructure(&self) -> Option<bool> {
        let mut out: std::os::raw::c_int = 0;
        let ok = unsafe {
            ffi::dnv_vista_sdk_gmod_node_metadata_install_substructure(self.as_ffi_ptr(), &mut out)
        };
        if ok != 0 {
            Some(out != 0)
        } else {
            None
        }
    }

    /// Number of normal assignment name entries.
    pub fn normal_assignment_name_count(&self) -> usize {
        unsafe {
            ffi::dnv_vista_sdk_gmod_node_metadata_normal_assignment_name_count(self.as_ffi_ptr())
        }
    }

    /// Normal assignment name `(key, value)` at `index`, or `None` if out of bounds.
    pub fn normal_assignment_name_at(&self, index: usize) -> Option<(&str, &str)> {
        let key_ptr = unsafe {
            ffi::dnv_vista_sdk_gmod_node_metadata_normal_assignment_name_key_at(
                self.as_ffi_ptr(),
                index,
            )
        };
        if key_ptr.is_null() {
            return None;
        }
        let value_ptr = unsafe {
            ffi::dnv_vista_sdk_gmod_node_metadata_normal_assignment_name_value_at(
                self.as_ffi_ptr(),
                index,
            )
        };
        assert!(
            !value_ptr.is_null(),
            "dnv_vista_sdk_gmod_node_metadata_normal_assignment_name_value_at returned NULL"
        );

        let key = unsafe { CStr::from_ptr(key_ptr) }
            .to_str()
            .expect("invalid UTF-8 in normal assignment name key");
        let value = unsafe { CStr::from_ptr(value_ptr) }
            .to_str()
            .expect("invalid UTF-8 in normal assignment name value");
        Some((key, value))
    }

    /// Iterator over all normal assignment name `(key, value)` pairs.
    pub fn normal_assignment_names(&self) -> impl Iterator<Item = (&str, &str)> {
        (0..self.normal_assignment_name_count()).map(move |i| {
            self.normal_assignment_name_at(i)
                .expect("index within count() must be valid")
        })
    }
}
